File: ocaml/examples/basic_example.ml
(* Copyright (c) Microsoft Corporation. All rights reserved. *)

(** Basic example: create a client, start a session, send a message. *)

open Copilot_sdk

let () =
  Lwt_main.run
    (let open Lwt.Syntax in
     (* Create client with default options *)
     let client = Client.create () in

     (* Start the client (connects to CLI) *)
     let* () = Client.start client in

     (* Check server status *)
     let* status = Client.get_status client in
     Printf.printf "Connected! Protocol version: %d\n" status.protocol_version;

     (* Create a session *)
     let config =
       { (Types.default_session_config ()) with
         model = Some "gpt-4o"
       }
     in
     let* session = Client.create_session ~config client in
     Printf.printf "Session created: %s\n" (Session.session_id session);

     (* Register an event handler *)
     Session.on session (fun event ->
       (match event.event_type with
        | Types.AssistantMessage ->
          let open Yojson.Safe.Util in
          let content =
            try event.data |> member "content" |> to_string
            with _ -> "(no content)"
          in
          Printf.printf "Assistant: %s\n" content
        | Types.AssistantMessageDelta ->
          let open Yojson.Safe.Util in
          let delta =
            try event.data |> member "content" |> to_string
            with _ -> ""
          in
          print_string delta
        | Types.SessionIdle ->
          Printf.printf "\n[Session idle]\n"
        | Types.SessionError ->
          let open Yojson.Safe.Util in
          let msg =
            try event.data |> member "message" |> to_string
            with _ -> "unknown error"
          in
          Printf.eprintf "Error: %s\n" msg
        | _ -> ());
       Lwt.return_unit);

     (* Register a permission handler that approves everything *)
     Session.register_permission_handler session Tools.approve_all;

     (* Send a message and wait for the response *)
     let msg = Types.make_message "What is OCaml? Explain in two sentences." in
     let* response = Session.send_and_wait session msg in
     (match response with
      | Some ev ->
        let open Yojson.Safe.Util in
        let content =
          try ev.data |> member "content" |> to_string
          with _ -> "(no content in final event)"
        in
        Printf.printf "\nFinal response: %s\n" content
      | None -> Printf.printf "Timed out waiting for response\n");

     (* Clean up *)
     let* () = Session.destroy session in
     let* () = Client.stop client in
     Printf.printf "Done!\n";
     Lwt.return_unit)

File: julia/examples/basic_example.jl
# --------------------------------------------------------------------------------------------
#  Copyright (c) Microsoft Corporation. All rights reserved.
# --------------------------------------------------------------------------------------------

# Basic example: connect to Copilot CLI, create a session, send a prompt, print the reply.

push!(LOAD_PATH, joinpath(@__DIR__, ".."))
using CopilotSDK

function main()
    # Create a client (set cli_path or COPILOT_CLI_PATH env var)
    client = CopilotClient(log_level=LOG_ERROR)

    try
        start!(client)
        println("Client connected.")

        # Create a session with automatic permission approval
        session = create_session(client;
            model="gpt-4",
            on_permission_request=approve_all,
        )
        println("Session created: $(session.session_id)")

        # Collect assistant messages
        done = Channel{Bool}(1)
        on(session) do event
            if event.type == "assistant.message_delta"
                content = get(event.data, "content", "")
                print(content)
            elseif event.type == "session.idle"
                println()
                try put!(done, true) catch; end
            end
        end

        # Send a prompt
        println("Sending prompt...")
        send(session, "Explain the Julia type system in three sentences.")

        # Wait for the response
        result = CopilotSDK.timedwait(60.0) do
            isready(done)
        end
        if result === :timed_out
            println("\nTimed out waiting for response.")
        end

        disconnect(session)
        println("Session disconnected.")
    finally
        stop!(client)
        println("Client stopped.")
    end
end

main()
